ui/search/nuxeo-search-form-layout.js

```diff
@@ -120,7 +120,8 @@ import { RoutingBehavior } from '../nuxeo-routing-behavior.js';
 
     _formHref(provider, searchName, hrefBase) {
       const name = (searchName || provider).toLowerCase();
-      const base = hrefBase || pathFromUrl(this.__dataHost.importPath || this.importPath);
+      const appImportPath = window.Nuxeo && window.Nuxeo.UI && window.Nuxeo.UI.app && window.Nuxeo.UI.app.importPath;
+      const base = hrefBase || pathFromUrl(appImportPath ? `${appImportPath}search/` : this.importPath);
```

Comment:
Here we are adding application specific logic to elements that should be as agnostic as possible:
1) the fact that Web UI reads the searches from the `search` folder
2) where to read the importPath from

These are not easy to fix, so I think we should discuss our options a bit before merging.

I can only think of a few options to fix 1):
a) make the `search` folder hardcoded (bad)
b) make the folder configurable via an `hrefTemplate` property just like nuxeo-document-layout, and allow it to also be read from a configuration property, so that Web UI can configure it without digging in the elements or us having to expose this property higher up (good?)

To fix 2), I can thing of a couple of options:
a) set the `hrefBase` everywhere we use it (bad)
b) allow nuxeo-search-form-layout.js  and nuxeo-search-results-layout.js to look for a parent with an importPath (not just the direct parent) and use it. (good?) Something along these lines (does not include fix for 1):
```diff
diff --git a/ui/search/nuxeo-search-results-layout.js b/ui/search/nuxeo-search-results-layout.js
index 7e747817..007bbbd7 100644
--- a/ui/search/nuxeo-search-results-layout.js
+++ b/ui/search/nuxeo-search-results-layout.js
@@ -102,9 +102,17 @@ import { I18nBehavior } from '../nuxeo-i18n-behavior.js';
       }
     }
 
+    get _parentImportPath() {
+      let host = this.__dataHost;
+      while (host && !host.importPath) {
+        host = host.__dataHost;
+      }
+      return host.importPath;
+    }
+
     _resultsHref(searchName, hrefBase) {
       const name = searchName.toLowerCase();
-      const base = hrefBase || pathFromUrl(this.__dataHost.importPath || this.importPath);
+      const base = hrefBase || pathFromUrl(this._parentImportPath || this.importPath);
       return `${base}${name}/${['nuxeo', name, 'search-results'].join('-')}.html`;
     }
```

Perhaps there's something more we can do here. Any ideas?

Reply:
In an ideal world elements outside of Web UI shouldn't expect any sort of convention.
If you look at document layouts, we have the exact same problem and the solution was having a `nuxeo-document-form-layout` where the application convention is enforced while still delegating the layout loading logics to a generic `nuxeo-layout`. 
I understand that with the picker we had to bring in `nuxeo-results-view` and thus all the convention based layout loading logics. Perhaps the better option would have been to introduce new elements and leave `nuxeo-search-results-layout` in Web UI enforcing its convention. 
That said, would it make sense to follow @Gabez0r 's suggestion and, at least when the data host is a TemplateInstance we went looking for it's owner (looking for the actual element that owns the template) and used its import path instead ? We know this isn't perfect but it is in line with the default values used in `nuxeo-layout` and the initial assumption that by default we'd be looking to load layouts relative to whoever is using them.

Reply:
note: in shadow dom I guess we could do `getRootNode().host` to reach the actual host (skipping all templates)

Reply:
> note: in shadow dom I guess we could do getRootNode().host to reach the actual host (skipping all templates)

This is a good suggestion, thanks. We still need to loop until we find an importPath, I believe.

> That said, would it make sense to follow @Gabez0r 's suggestion and, at least when the data host is a TemplateInstance we went looking for it's owner (looking for the actual element that owns the template) and used its import path instead ? We know this isn't perfect but it is in line with the default values used in nuxeo-layout and the initial assumption that by default we'd be looking to load layouts relative to whoever is using them.

@nelsonsilva: This makes sense to me, but how do we solve the issue about the fact the layouts are inside the `search` folder, given that this is not reflected on the importPath? By creating an element on Web UI that wraps nuxeo-search-results-layout?

Reply:
>2)b) allow nuxeo-search-form-layout.js and nuxeo-search-results-layout.js to look for a parent with an importPath (not just the direct parent) and use it.

How is that different from the other solution that I mentioned above?

>One other solution that was explored was to recursively check for the `__dataHost` (instead of only checking the direct one), but it's difficult to find a stopping condition that makes sense to know whether or not a Web UI element has been reached. One could assume that we want to go all the way back to `<nuxeo-app>` through this method, but this ends up being problematic in the context of slots, as you'll end up reaching the `<html>` node of the DOM before getting to `<nuxeo-app>`.

Reply:
>If you look at document layouts, we have the exact same problem and the solution was having a `nuxeo-document-form-layout` where the application convention is enforced while still delegating the layout loading logics to a generic `nuxeo-layout`.

It's not the same problem because `<nuxeo-document-form-layout>` uses `<nuxeo-document-layout>`, an ELEMENTS element that can afford to [rely on its direct parent to get the right `importPath`](https://github.com/nuxeo/nuxeo-elements/blob/master/ui/nuxeo-document-layout.js#L242) because it is always used by Web UI elements. If `<nuxeo-document-layout>` were to be used by another ELEMENTS element instead, then you'd have the exact same problem as the search layouts.

Reply:
>> 2)b) allow nuxeo-search-form-layout.js and nuxeo-search-results-layout.js to look for a parent with an importPath (not just the direct parent) and use it.

> How is that different from the other solution that I mentioned above?

It's different in the sense that we do not introduce a dependency on the nuxeo app, which is only one of the potential clients the nuxeo-elements can have.  So instead of using `window.Nuxeo.UI.app.importPath` we would use a getter (_parentImportPath) to find an importPath which can come from nuxeo-app or not.

Reply:
>It's different in the sense that we do not introduce a dependency on the nuxeo app, which is only one of the potential clients the nuxeo-elements can have. So instead of using window.Nuxeo.UI.app.importPath we would use a getter (_parentImportPath) to find an importPath which can come from nuxeo-app or not.

The only solution that strictly depends on the `<nuxeo-app>` is the "Proposed Solution". However, I've described another approach (above, under "Other Solutions") that does, from what I understand, what you are suggesting. I've explained why that approach is problematic:
1. Finding a parent `importPath` does not mean finding the correct `importPath`. If the parent element is an Elements element, this path will be incorrect. The problem is that we don't have a proper stop condition, i.e. we cannot properly distinguish an Elements `importPath` from a "main app" `importPath` (whether it is from Web UI or something else).
1. In the context of slots we might even get situations in which recursively checking for a parent `importPath` will never return a "main app" `importPath` (whether it is from Web UI or something else).